### ISSUE TITLE
Java8 type-use annotations on arrays not supported

### ISSUE THREAD

**Issue by metlos** [checkstyle/checkstyle#904]
Checkstyle fails to parse constructs like:

```java
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import java.util.Set;

public class CheckstyleBugs {
    private String @Fancy [] array; // !!!
    Set<String @Fancy []> set;       // !!!
}

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE_USE)
@interface Fancy {
}

```
giving "unexpected token" errors.

**Comment by romani** [comment]
https://docs.oracle.com/javase/specs/jls/se8/html/jls-10.html#jls-10.2

>
```
Because of how array types are formed, the following parameter declarations have the same array type:

void m(int @A [] @B []  x) {}
void n(int @A [] @B ... y) {}

And perhaps surprisingly, the following field declarations have the same array type:

int @A [] f @B [];
int @B [] @A [] g;
```

**Comment by romani** [comment]
@metlos , please provide compilable code to let quickly reproduce a problem:

```
$ cat CheckstyleBugs.java 
import javax.annotation.Nullable;

public class CheckstyleBugs<T> {

    String @Nullable [] array;
}


$ javac -cp jsr305-1.3.9.jar CheckstyleBugs.java 
CheckstyleBugs.java:5: error: annotation type not applicable to this kind of declaration
    String @Nullable [] array;
           ^
1 error

$ java -version
java version "1.8.0_25"
Java(TM) SE Runtime Environment (build 1.8.0_25-b17)
Java HotSpot(TM) 64-Bit Server VM (build 25.25-b02, mixed mode)

```

**Comment by metlos** [comment]
Oh, sorry. The `@Nullable` I had in mind in my example was the one from the checkerframework, not the one in javax.annotation. Apologies for not stating it explicitly.

A simple example code might look like this:
```java
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

public class CheckstyleBugs {
    private String @Fancy [] array;
}

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE_USE)
@interface Fancy {
}
```

**Comment by romani** [comment]
workaround is : 
```java
    private @Fancy  String [] array;
    Set<@Fancy  String []> set;
```

@metlos , how blocking is this problem for you ?

**Comment by metlos** [comment]
That is unfortunately not a workaround.

Let me explain the semantic difference on the example of @NotNull on which the difference is more easily explained:

```
@NotNull String[]
```
is an array of strings where all the strings are not null.

```
String @NotNull []
```
Is an array of strings, where the array itself is never null.

This problem basically disallows people to use both checkerframework and checkstyle in one project if they use arrays in their code.

**Comment by romani** [comment]
@metlos , thanks for explanation, I marking issue as "hight demand".

**Comment by romani** [comment]
fixed by #2891

**Comment by romani** [comment]
@metlos  , 
as soon as https://travis-ci.org/checkstyle/checkstyle/jobs/108286542 is finished , binaries will be at 
https://oss.sonatype.org/content/repositories/snapshots/com/puppycrawl/tools/checkstyle/6.16-SNAPSHOT/ . Please test your sources to make sure that did not miss anything. Feel free to open new issues.
